### PR TITLE
Add residuals for quadratic nllhs with parameter dependent sigma

### PR DESCRIPTION
**Pull request by FFroehlich**
default `min_sigma_ = 50` corresponds to d2d default. This enables optimization of respective objective functions using least squares optimizers, but also allows changing the FIM accordingly, which is then used in trust region optimizers.

### CODE REVIEW

File: src/rdata.cpp [hunk 1]
```diff
@@ -833,6 +854,14 @@ void ReturnData::fsres(const int it, Model &model, const ExpData &edata) {
             sres.at(idx) = amici::fsres(y_it.at(iy), sy_it.at(iy + ny * ip),
                                         observedData[iy], sigmay_it.at(iy),
                                         ssigmay_it.at(iy + ny * ip));
+            if (sigma_res) {
+                int idx_res =
+                    (iy + it * edata.nytrue() + edata.nytrue() * edata.nt()) *
```
Comment (dweindl) [inline src/rdata.cpp:859]: cache result of `edata.n*()` outside nested loop? Could also avoid some repeated `.at()`s, but not sure how relevant.

**Review by dweindl — APPROVED** [review]

**Review by yannikschaelte — APPROVED** [review]
the formulas are correct (one might want to document how it works, i.e. introduction of additional residuals), aside nothing except what daniel pointed out

File: include/amici/model.h [hunk 2]
```diff
@@ -1704,6 +1740,12 @@ class Model : public AbstractModel, public ModelDimensions {
      * checked for finiteness
      */
     bool always_check_finite_ {false};
+    
+    /** indicates whether sigma residuals are to be added for every  */
+    bool sigma_res_ {false};
+    
+    /** offset to ensure positivity of sigma residuals, only has an effect when `sigma_res_` is `true`  */
+    realtype min_sigma_ {50.0};
```
Comment (paulstapor) [inline include/amici/model.h:1748]: Just as a remark:
The higher the offset, the weaker the contribution of this term to the gradient will be, right? The square root function flattens out for higher values... So, I my gut feeling tells me that this offset term influences the parameter estimation process somehow. Therefore, I would suggest a lower value than 50, which is however still high enough to not cause problems for typical applications. I think already 10 should do the job here...

Comment (yannikschaelte) [reply include/amici/model.h:1748]: I agree that the offset should have an influence. No idea about practically typical values (ok, log(2*pi*1e-8^2)=-35, one could look at typical bounds in the benchmark collection).

File: src/rdata.cpp (hunk 1 above)
Comment (FFroehlich) [reply src/rdata.cpp:859]: as those are not called during simulation, computational cost is pretty negligible here.

File: include/amici/model.h (hunk 2 above)
Comment (FFroehlich) [reply include/amici/model.h:1749]: Note that we do not account for the `2pi` factor here (we could). The gradient itself is not influenced by C, but `sres` is and the FIM is. Probably the smartest way to go about this is set some value based on bounds for sigmas in pypesto.

**Pull request merged**